fdbserver/storageserver.actor.cpp

```diff
@@ -5875,6 +5877,23 @@ ACTOR Future<std::unordered_map<Key, Version>> dispatchChangeFeeds(StorageServer
 	}
 }
 
+namespace {
+ACTOR Future<Version> getRawReadVersion(Transaction* pTransaction) {
+	loop {
+		try {
+			Version readVersion = wait(pTransaction->getRawReadVersion());
+			return readVersion;
+		} catch (Error& err) {
+			if (err.code() != error_code_proxy_memory_limit_exceeded &&
+			    err.code() != error_code_batch_transaction_throttled) {
+				throw err;
```

Comment:
I don't think this should throw any errors.

Any errors returned by the GRV (except cancelled error) needs to be consumed inside the FetchKey without terminating the SS.

Reply:
This function should only throw `error_code_actor_cancelled`.

Reply:
Yes.

Reply:
Maybe we don't need this function, but call `pTransaction->getRawReadVersion()` in fetchKeys, and if it fails, the error is assigned to lastError, and do the error handling at [a single place](https://github.com/apple/foundationdb/blob/83a317ef888e2e1561b593eaea109a242aff1c2a/fdbserver/storageserver.actor.cpp#L6276).

Reply:
That place picks which error to retry. I think it is vulnerable because GRV may implement returning new error in the future, which may cause SS to terminate. Instead, maybe we should pick which error to throw, and for all other errors, let's retry.

Reply:
> Maybe we don't need this function, but call `pTransaction->getRawReadVersion()` in fetchKeys, and if it fails, the error is assigned to lastError, and do the error handling at [a single place](https://github.com/apple/foundationdb/blob/83a317ef888e2e1561b593eaea109a242aff1c2a/fdbserver/storageserver.actor.cpp#L6276).

That catch does not capture exceptions for `tr->getRawReadVersion()`. This call is outside the corresponding try block.

Reply:
> That place picks which error to retry. I think it is vulnerable because GRV may implement returning new error in the future, which may cause SS to terminate. Instead, maybe we should pick which error to throw, and for all other errors, let's retry.

I think the `Transaction::onError` should handle the errors instead.